py/frequenz/client/electricity_trading/_types.py

```diff
@@ -0,0 +1,1220 @@
+            delivery_period=DeliveryPeriod.from_pb(order.delivery_period),
+            type=OrderType.from_pb(order.type),
+            side=MarketSide.from_pb(order.side),
+            price=Price.from_pb(order.price),
+            quantity=Energy.from_pb(order.quantity),
+            stop_price=Price.from_pb(order.stop_price) if order.stop_price else None,
+            peak_price_delta=Price.from_pb(order.peak_price_delta)
+            if order.peak_price_delta
+            else None,
+            display_quantity=Energy.from_pb(order.display_quantity)
+            if order.display_quantity
+            else None,
+            execution_option=OrderExecutionOption.from_pb(order.execution_option)
+            if order.execution_option
+            else None,
+            valid_until=order.valid_until.ToDatetime() if order.valid_until else None,
+            payload=json_format.MessageToDict(order.payload) if order.payload else None,
+            tag=order.tag if order.tag else None,
+        )
+
+    def to_pb(self) -> electricity_trading_pb2.Order:
+        """
+        Convert an Order object to protobuf Order.
+
+        Returns:
+            Protobuf message corresponding to the Order object.
+        """
+        if self.valid_until:
+            valid_until = timestamp_pb2.Timestamp()
+            valid_until.FromDatetime(self.valid_until)
+        else:
+            valid_until = None
+        return electricity_trading_pb2.Order(
+            delivery_area=self.delivery_area.to_pb(),
+            delivery_period=self.delivery_period.to_pb(),
+            type=electricity_trading_pb2.OrderType.ValueType(self.type.value),
+            side=electricity_trading_pb2.MarketSide.ValueType(self.side.value),
+            price=self.price.to_pb(),
+            quantity=self.quantity.to_pb(),
+            stop_price=self.stop_price.to_pb() if self.stop_price else None,
+            peak_price_delta=self.peak_price_delta.to_pb()
+            if self.peak_price_delta
+            else None,
+            display_quantity=self.display_quantity.to_pb()
+            if self.display_quantity
+            else None,
+            execution_option=(
+                electricity_trading_pb2.OrderExecutionOption.ValueType(
+                    self.execution_option.value
+                )
+                if self.execution_option
+                else None
+            ),
+            valid_until=valid_until,
+            payload=struct_pb2.Struct(fields=self.payload) if self.payload else None,
+            tag=self.tag if self.tag else None,
+        )
+
+
+@dataclass(frozen=True)
+class StateDetail:
+    """
+    Details about the current state of the order.
+
+    Attributes:
+        state: Current state of the order.
+        state_reason: Reason for the current state.
+        market_actor: Actor responsible for the current state.
+    """
+
+    state: OrderState
+    state_reason: StateReason
+    market_actor: MarketActor
```

Comment:
Normally, something like below is better for LSP and for the documentation site, but updating is a lot of work, so maybe you just want to make an issue for now.  I can make a script to do that as well, if that helps.

```suggestion
class StateDetail:
    """Details about the current state of the order."""

    state: OrderState
    """Current state of the order."""
    
    state_reason: StateReason
    """Reason for the current state."""
    
    market_actor: MarketActor
    """Actor responsible for the current state."""
```

Reply:
Ah fair enough that's how I started to do it but then I saw that the weather API was done differently so I changed it. I can change it back though

Reply:
oops :D